### PR TITLE
Support more TARGET_VENDOR values

### PR DESCRIPTION
**Pull request by shr-project**
Only a draft, because for some reason this still doesn't work correctly and compiler-rt is failing with:
http://errors.yoctoproject.org/Errors/Details/574365/

### CODE REVIEW

**Comment by kraj** [comment]
I think this will be more error prone and  its heavily editing sourcefiles which can silently go wrong on clang upgrades. I am not too excited about this patchset. While. patcing via patch is explicit and forces to do the right thing as patch may not apply cleanly when code moves around or get re-implemented, which happens all the time in the clang driver in this area.

**Comment by shr-project** [comment]
so should I send PR to add webos, luneos to these patches or what makes poky, yoe, wrs special to have this support OOTB.

with CLANG_EXTRA_VENDORS there is at only 1 place we need to update when upgrading to newer clang/llvm while with each DISTRO adding patches on top of patches everybody will need to update when there is a conflict.

I can also add an extra grep before sed to check that expected pattern is still there or show the diff caused by sed with bb.warn or something like that to make it a bit more visibile when it stops applying.

Also this behavior is new in master branch, with gatesgarth I was able to build compiler-rt fine with -webos in TARGET_VENDOR, do you remember what changed which requires this explicit list of supported tripples now?

**Comment by shr-project** [comment]
Or probably safer solution would be for patches to add some placeholder string after the oe variant and then the sed call just replacing the place holder?

**Comment by shr-project** [comment]
Maybe bbwarn is too much as it's shown for each recipe using common.inc:

```
WARNING: clang-native-12.0.0-r0 do_configure: Adding support following TARGET_VENDOR values: poky yoe wrs in /OE/build/oe-core/tmp-glibc/work-shared/llvm-project-source-12.0.0-r0/git/llvm/lib/Support/Triple.cpp and /OE/build/oe-core/tmp-glibc/work-shared/llvm-project-source-12.0.0-r0/git/clang/lib/Driver/ToolChains/Gnu.cpp
WARNING: clang-cross-x86_64-12.0.0-r0 do_configure: Adding support following TARGET_VENDOR values: poky yoe wrs in /OE/build/oe-core/tmp-glibc/work-shared/llvm-project-source-12.0.0-r0/git/llvm/lib/Support/Triple.cpp and /OE/build/oe-core/tmp-glibc/work-shared/llvm-project-source-12.0.0-r0/git/clang/lib/Driver/ToolChains/Gnu.cpp
WARNING: compiler-rt-12.0.0-r0 do_configure: Adding support following TARGET_VENDOR values: poky yoe wrs in /OE/build/oe-core/tmp-glibc/work-shared/llvm-project-source-12.0.0-r0/git/llvm/lib/Support/Triple.cpp and /OE/build/oe-core/tmp-glibc/work-shared/llvm-project-source-12.0.0-r0/git/clang/lib/Driver/ToolChains/Gnu.cpp
```

Will change it to just bbnote.

**Comment by shr-project** [comment]
Moving this to llvm-project-source do_patch postfunc allows to do this only once and also allows to use TARGET_VENDOR as it's not BUILD_VENDOR for  llvm-project-source (like it was for clang-native).

File: recipes-devtools/clang/clang/0026-For-x86_64-set-Yocto-based-GCC-install-search-path.patch [hunk 1]
```diff
@@ -59,7 +64,7 @@ index 05d1d3003881..e56812145e6d 100644
        "x86_64-redhat-linux",    "x86_64-suse-linux",
        "x86_64-manbo-linux-gnu", "x86_64-linux-gnu",
        "x86_64-slackware-linux", "x86_64-unknown-linux",
-+      "x86_64-oe-linux", "x86_64-poky-linux", "x86_64-wrs-linux",
++      "x86_64-oe-linux", CLANG_EXTRA_OE_VENDORS_TRIPLES
```
Comment (kraj) [inline recipes-devtools/clang/clang/0026-For-x86_64-set-Yocto-based-GCC-install-search-path.patch:67]: Lets make it `//CLANG_EXTRA_OE_VENDORS_TRIPLES` so the code can keep compiling outside OE builds.

File: recipes-devtools/clang/clang/0029-llvm-Recognize-yoe-and-poky-as-OE-distro.patch [hunk 2]
```diff
@@ -17,13 +22,11 @@ diff --git a/llvm/lib/Support/Triple.cpp b/llvm/lib/Support/Triple.cpp
...
-@@ -490,6 +490,9 @@ static Triple::VendorType parseVendor(StringRef VendorName) {
+@@ -490,6 +490,6 @@ static Triple::VendorType parseVendor(StringRef VendorName) {
      .Case("mesa", Triple::Mesa)
      .Case("suse", Triple::SUSE)
-     .Case("oe", Triple::OpenEmbedded)
-+    .Case("yoe", Triple::OpenEmbedded)
-+    .Case("poky", Triple::OpenEmbedded)
-+    .Case("wrs", Triple::OpenEmbedded)
+-    .Case("oe", Triple::OpenEmbedded)
++    .Case("oe", Triple::OpenEmbedded)CLANG_EXTRA_OE_VENDORS_CASES
```
Comment (kraj) [inline recipes-devtools/clang/clang/0029-llvm-Recognize-yoe-and-poky-as-OE-distro.patch:29]: Lets make it `//CLANG_EXTRA_OE_VENDORS_CASES` so the code can keep compiling outside OE builds.

File: recipes-devtools/clang/llvm-project-source.inc [hunk 3]
```diff
@@ -18,3 +18,20 @@ STAMPCLEAN = "${STAMPS_DIR}/work-shared/llvm-project-source-${PV}-*"
...
+add_more_target_vendors() {
+    local cases="" triples=""
+    for dash_vendor in ${CLANG_EXTRA_OE_VENDORS}; do
+        vendor=`echo $dash_vendor | sed 's/^-//g'`
+        cases="$cases.Case(\"$vendor\", Triple::OpenEmbedded)"
+        triples="$triples\"x86_64-$vendor-linux\","
+    done
+    bbnote "Adding support following TARGET_VENDOR values: ${CLANG_EXTRA_OE_VENDORS} in ${S}/llvm/lib/Support/Triple.cpp and ${S}/clang/lib/Driver/ToolChains/Gnu.cpp"
+    sed "s/CLANG_EXTRA_OE_VENDORS_CASES/$cases/g" -i ${S}/llvm/lib/Support/Triple.cpp
+    sed "s/CLANG_EXTRA_OE_VENDORS_TRIPLES/$triples/g" -i ${S}/clang/lib/Driver/ToolChains/Gnu.cpp
```
Comment (kraj) [inline recipes-devtools/clang/llvm-project-source.inc:34]: replace `//CLANG_EXTRA_OE_VENDORS_CASES` and `//CLANG_EXTRA_OE_VENDORS_TRIPLES` here

**Review by kraj — CHANGES_REQUESTED** [review]

**Comment by kraj** [comment]
> Moving this to llvm-project-source do_patch postfunc allows to do this only once and also allows to use TARGET_VENDOR as it's not BUILD_VENDOR for llvm-project-source (like it was for clang-native).

yes doing it as part of llvm-project-source might be right thing but please check it works for native/nativesdk/target cases and signatures don't change between them

**Review by kraj — APPROVED** [review]
lgtm

**Comment by kraj** [comment]
@shr-project while I merged it, its causing a problem where clang-native is compiling for everytime I chage TCLIBC or MACHINE which is not expected, common sources are prepared once and they are not supposed to change signatures. clang-native is even reused between musl/glibc/machine transitions unlike gcc where compiler is compiled everytime these vars change.

To reproduce it


TCLIBC="musl" bitbake clang
TCLIBC="glibc" bitbake clang

and it ends up recompiling clang native in second run above which was not case before

**Comment by kraj** [comment]
here is bitbake diffsigs

```
% bitbake-diffsigs /mnt/b/yoe/master/build/tmp/stamps/work-shared/llvm-project-source-12.0.0-r0.do_patch.sigdata.c290977dbf9af091c0cefca06e2b7a626b3250db7a999653b99120d6db31da6c /mnt/b/yoe/master/build/tmp/sta
mps/work-shared/llvm-project-source-12.0.0-r0.do_patch.sigdata.ed254a989913203c7934a21e7e45e51a23648f22e7b3203cd17308411308b671 
NOTE: Starting bitbake server...                                                                        
NOTE: Started PRServer with DBfile: /mnt/b/yoe/master/cache/prserv.sqlite3, IP: 127.0.0.1, PORT: 46153, PID: 2013765
Task dependencies changed from:                                                                                                                                                                                  
['ABIEXTENSION', 'AR', 'AS', 'ASNEEDED', 'BUILD_AR', 'BUILD_AS', 'BUILD_AS_ARCH', 'BUILD_CC', 'BUILD_CCLD', 'BUILD_CC_ARCH', 'BUILD_CFLAGS', 'BUILD_CPP', 'BUILD_CPPFLAGS', 'BUILD_CXX', 'BUILD_CXXFLAGS', 'BUILD
_FC', 'BUILD_LD', 'BUILD_LDFLAGS', 'BUILD_LD_ARCH', 'BUILD_NM', 'BUILD_OPTIMIZATION', 'BUILD_PREFIX', 'BUILD_RANLIB', 'BUILD_STRIP', 'CC', 'CCLD', 'CFLAGS', 'CLANG_EXTRA_OE_VENDORS', 'CLANG_TIDY_toolchain-clan
g', 'CPP', 'CPPFLAGS', 'CXX', 'CXXFLAGS', 'DEBUG_BUILD', 'DEBUG_FLAGS', 'DEBUG_OPTIMIZATION', 'DEBUG_PREFIX_MAP', 'DEFAULTTUNE', 'EXTENDPE', 'FC', 'FULL_OPTIMIZATION', 'GCCPIE', 'HOST_ARCH', 'HOST_AS_ARCH', 'H
OST_CC_ARCH', 'HOST_LD_ARCH', 'HOST_OS', 'HOST_PREFIX', 'HOST_SYS', 'HOST_VENDOR', 'LC_ALL', 'LD', 'LDFLAGS', 'LIBCEXTENSION', 'LIBCPLUSPLUS', 'LINKER_HASH_STYLE', 'LOGFIFO', 'MAKE', 'NM', 'OBJCOPY', 'OBJDUMP'
, 'OPTLEVEL', 'P', 'PATCHTOOL', 'PE', 'PIGZ', 'PKG_CONFIG_DIR', 'PKG_CONFIG_DISABLE_UNINSTALLED', 'PKG_CONFIG_LIBDIR', 'PKG_CONFIG_PATH', 'PKG_CONFIG_SYSROOT_DIR', 'PKG_CONFIG_SYSTEM_INCLUDE_PATH', 'PKG_CONFIG
_SYSTEM_LIBRARY_PATH', 'PN', 'PR', 'PSEUDO_DISABLED', 'PV', 'QA_LOGFILE', 'RANLIB', 'READELF', 'RECIPE_SYSROOT_NATIVE', 'S', 'SECURITY_CFLAGS', 'SECURITY_LDFLAGS', 'SECURITY_PIE_CFLAGS', 'SECURITY_STACK_PROTEC
TOR', 'SECURITY_STRINGFORMAT', 'SELECTED_OPTIMIZATION', 'STAGING_BASE_LIBDIR_NATIVE', 'STAGING_DATADIR', 'STAGING_DIR_NATIVE', 'STAGING_INCDIR_NATIVE', 'STAGING_LIBDIR_NATIVE', 'STRINGS', 'STRIP', 'T', 'TARGET
_ARCH', 'TARGET_AS_ARCH', 'TARGET_CC_ARCH', 'TARGET_CFLAGS', 'TARGET_CPPFLAGS', 'TARGET_CXXFLAGS', 'TARGET_LDFLAGS', 'TARGET_LD_ARCH', 'TARGET_LINK_HASH_STYLE', 'TARGET_OS', 'TARGET_PREFIX', 'TARGET_SYS', 'TAR
GET_VENDOR', 'TOOLCHAIN_OPTIONS', 'TUNE_ARCH', 'TUNE_ARCH_tune-riscv64', 'TUNE_ASARGS', 'TUNE_CCARGS', 'TUNE_LDARGS', 'TZ', 'add_more_target_vendors', 'base_bindir', 'base_libdir', 'base_libdir_native', 'base_
prefix', 'base_sbindir', 'baselib', 'bbnote', 'bindir', 'datadir', 'do_qa_patch', 'docdir', 'exec_prefix', 'includedir', 'includedir_native', 'infodir', 'lcl_maybe_fortify', 'libdir', 'libdir_native', 'libexec
dir', 'localstatedir', 'mandir', 'nonarch_base_libdir', 'nonarch_libdir', 'oldincludedir', 'package_qa_handle_error', 'package_qa_write_error', 'patch_do_patch', 'prefix', 'prefix_native', 'root_prefix', 'sbin
dir', 'servicedir', 'sharedstatedir', 'src_patches', 'sysconfdir', 'systemd_system_unitdir', 'systemd_unitdir', 'systemd_user_unitdir']
to:                                                                                                     
['ABIEXTENSION', 'AR', 'AS', 'ASNEEDED', 'BUILD_AR', 'BUILD_AS', 'BUILD_AS_ARCH', 'BUILD_CC', 'BUILD_CCLD', 'BUILD_CC_ARCH', 'BUILD_CFLAGS', 'BUILD_CPP', 'BUILD_CPPFLAGS', 'BUILD_CXX', 'BUILD_CXXFLAGS', 'BUILD
_FC', 'BUILD_LD', 'BUILD_LDFLAGS', 'BUILD_LD_ARCH', 'BUILD_NM', 'BUILD_OPTIMIZATION', 'BUILD_PREFIX', 'BUILD_RANLIB', 'BUILD_STRIP', 'CC', 'CCLD', 'CFLAGS', 'CLANG_EXTRA_OE_VENDORS', 'CLANG_TIDY_toolchain-clan
g', 'CPP', 'CPPFLAGS', 'CXX', 'CXXFLAGS', 'DEBUG_BUILD', 'DEBUG_FLAGS', 'DEBUG_OPTIMIZATION', 'DEBUG_PREFIX_MAP', 'DEFAULTTUNE', 'EXTENDPE', 'FC', 'FULL_OPTIMIZATION', 'GCCPIE', 'HOST_ARCH', 'HOST_AS_ARCH', 'H
OST_CC_ARCH', 'HOST_LD_ARCH', 'HOST_OS', 'HOST_PREFIX', 'HOST_SYS', 'HOST_VENDOR', 'LC_ALL', 'LD', 'LDFLAGS', 'LIBCEXTENSION', 'LIBCPLUSPLUS', 'LINKER_HASH_STYLE', 'LOGFIFO', 'MAKE', 'NM', 'OBJCOPY', 'OBJDUMP'
, 'OPTLEVEL', 'P', 'PATCHTOOL', 'PE', 'PIGZ', 'PKG_CONFIG_DIR', 'PKG_CONFIG_DISABLE_UNINSTALLED', 'PKG_CONFIG_LIBDIR', 'PKG_CONFIG_PATH', 'PKG_CONFIG_SYSROOT_DIR', 'PKG_CONFIG_SYSTEM_INCLUDE_PATH', 'PKG_CONFIG
_SYSTEM_LIBRARY_PATH', 'PN', 'PR', 'PSEUDO_DISABLED', 'PV', 'QA_LOGFILE', 'RANLIB', 'READELF', 'RECIPE_SYSROOT_NATIVE', 'S', 'SECURITY_CFLAGS', 'SECURITY_LDFLAGS', 'SECURITY_PIE_CFLAGS', 'SECURITY_STACK_PROTEC
TOR', 'SECURITY_STRINGFORMAT', 'SELECTED_OPTIMIZATION', 'STAGING_BASE_LIBDIR_NATIVE', 'STAGING_DATADIR', 'STAGING_DIR_NATIVE', 'STAGING_INCDIR_NATIVE', 'STAGING_LIBDIR_NATIVE', 'STRINGS', 'STRIP', 'T', 'TARGET
_ARCH', 'TARGET_AS_ARCH', 'TARGET_CC_ARCH', 'TARGET_CFLAGS', 'TARGET_CPPFLAGS', 'TARGET_CXXFLAGS', 'TARGET_LDFLAGS', 'TARGET_LD_ARCH', 'TARGET_LINK_HASH_STYLE', 'TARGET_OS', 'TARGET_PREFIX', 'TARGET_SYS', 'TAR
GET_VENDOR', 'TOOLCHAIN_OPTIONS', 'TUNE_ARCH', 'TUNE_ARCH_tune-riscv32', 'TUNE_ASARGS', 'TUNE_CCARGS', 'TUNE_LDARGS', 'TZ', 'add_more_target_vendors', 'base_bindir', 'base_libdir', 'base_libdir_native', 'base_
prefix', 'base_sbindir', 'baselib', 'bbnote', 'bindir', 'datadir', 'do_qa_patch', 'docdir', 'exec_prefix', 'includedir', 'includedir_native', 'infodir', 'lcl_maybe_fortify', 'libdir', 'libdir_native', 'libexec
dir', 'localstatedir', 'mandir', 'nonarch_base_libdir', 'nonarch_libdir', 'oldincludedir', 'package_qa_handle_error', 'package_qa_write_error', 'patch_do_patch', 'prefix', 'prefix_native', 'root_prefix', 'sbin
dir', 'servicedir', 'sharedstatedir', 'src_patches', 'sysconfdir', 'systemd_system_unitdir', 'systemd_unitdir', 'systemd_user_unitdir']
basehash changed from 40a2d531bb532c4f6f5ebbbdc95502b29a6ff2ac88bac98c89d8b7ad8a317e73 to 070902e28832146f6d51b9462bc3f926ed266780b8350c987a2e4efae3c95d86
List of dependencies for variable LIBCEXTENSION changed from 'set()' to '{'ABIEXTENSION'}'
List of dependencies for variable TUNE_ARCH changed from '{'TUNE_ARCH_tune-riscv64', 'DEFAULTTUNE'}' to '{'TUNE_ARCH_tune-riscv32', 'DEFAULTTUNE'}'
changed items: {'TUNE_ARCH_tune-riscv32', 'TUNE_ARCH_tune-riscv64'}
Dependency on variable TUNE_ARCH_tune-riscv32 was added
Dependency on Variable TUNE_ARCH_tune-riscv64 was removed
Variable LIBCEXTENSION value changed:
"[--musl-] {+${@['', '-gnu'][(d.getVar('ABIEXTENSION') or '') != '']}+}"
Variable add_more_target_vendors value changed:
@@ -1,7 +1,7 @@
     local cases="" triples=""
     for dash_vendor in ${CLANG_EXTRA_OE_VENDORS}; do
         vendor=`echo $dash_vendor | sed 's/^-//g'`
-        cases="$cases\n    .Case(\"$vendor\", Triple::OpenEmbedded)"
+        cases="$cases.Case(\"$vendor\", Triple::OpenEmbedded)"
         triples="$triples\"x86_64-$vendor-linux\"," 
     done
     bbnote "Adding support following TARGET_VENDOR values: ${CLANG_EXTRA_OE_VENDORS} in ${S}/llvm/lib/Support/Triple.cpp and ${S}/clang/lib/Driver/ToolChains/Gnu.cpp"
Variable DEFAULTTUNE value changed from 'riscv64' to 'riscv32'
````

**Comment by kraj** [comment]
I think this is a long tail. `add_more_target_vendors` should be converted to a python function

**Pull request merged**